src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -418,117 +255,56 @@ private Type EmitCoercionCode(AssociativeNode arg, Type argType, Type param)
-            EmitOpCode(OpCodes.Ldloc, currentVarIndex);
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-
-            if (typeof(Source) == typeof(double))
-            {
-                EmitOpCode(OpCodes.Ldelem_R8);
-            }
-            else if (typeof(Source) == typeof(long))
-            {
-                EmitOpCode(OpCodes.Ldelem_I8);
-            }
-            else if (typeof(Source) == typeof(int))
+            EmitOpCode(OpCodes.Ldnull);
+            MethodInfo convertMethod = null;
+            if (typeof(Target) == typeof(int))
             {
-                EmitOpCode(OpCodes.Ldelem_I4);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToInt32), new[] { typeof(Source) });
             }
-            if (typeof(Target) == typeof(int))
+            else if(typeof(Target) == typeof(double))
             {
-                if (typeof(Source) == typeof(double))
-                {
-                    EmitMathRound();
-                }
-                EmitOpCode(OpCodes.Conv_I4);
-                EmitOpCode(OpCodes.Stelem_I4);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToDouble), new[] { typeof(Source) });
             }
             else if (typeof(Target) == typeof(long))
             {
-                if (typeof(Source) == typeof(double))
-                {
-                    EmitMathRound();
-                }
-                EmitOpCode(OpCodes.Conv_I8);
-                EmitOpCode(OpCodes.Stelem_I8);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToInt64), new[] { typeof(Source) });
             }
-            if (typeof(Target) == typeof(double))
+            else if (typeof(Target) == typeof(bool))
             {
-                EmitOpCode(OpCodes.Conv_R8);
-                EmitOpCode(OpCodes.Stelem_R8);
+                convertMethod = typeof(Convert).GetMethod(nameof(Convert.ToBoolean), new[] { typeof(Source) });
             }
-            // i++;
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
-            EmitOpCode(OpCodes.Ldc_I4_1);
+            EmitOpCode(OpCodes.Ldftn, convertMethod);
 
-            EmitOpCode(OpCodes.Add);
+            var ctor = System.Linq.Expressions.Expression.GetDelegateType(typeof(Source), typeof(Target))
+                .GetConstructors()[0];
+            EmitOpCode(OpCodes.Newobj, ctor);
             
-            EmitOpCode(OpCodes.Stloc, counterIndex);
-
-            // i < arr.Length;
-            MarkLabel(loopCondLabel.Value,"label:cond");
-
-            EmitOpCode(OpCodes.Ldloc, counterIndex);
+            var selectMethod = typeof(Enumerable).GetMember(nameof(Enumerable.Select)).OfType<MethodInfo>().First();
 
-            // Load input array
-            EmitOpCode(OpCodes.Ldloc, currentVarIndex);
-            EmitOpCode(OpCodes.Ldlen);
-            EmitOpCode(OpCodes.Conv_I4);
+            var genericSelect = selectMethod.MakeGenericMethod(typeof(Source), typeof(Target));
```

Comment:
if we find that something like this is slow, we could likely precompute many of the common versions.